### PR TITLE
Add build function for custom initialization

### PR DESCRIPTION
**Pull request by dsteurer**
Would you consider allowing a custom function to initialize keys instead of the global init-key multimethod? (See the attached commit.) A custom function could be useful for example for wrapping key initialization with code for logging, profiling, error handling, or asynchronous initialization.

### CODE REVIEW

**Comment by weavejester** [comment]
I've been considering this, but rather than bundling it in `init`, I was thinking of having a separate function. I'm just not sure what to call it, yet. Maybe `update` or `update-and-expand`.

**Comment by weavejester** [comment]
What do you think of calling the function `build`?

**Comment by levand** [comment]
+1 on a separate function, and defining `init` and `halt` in terms of it. That's basically what Component does (https://github.com/stuartsierra/component/blob/master/src/com/stuartsierra/component.cljc#L129) , and it's come in handy so many times for performing arbitrary operations in dependency order. It makes the whole system surprisingly general and powerful.

**Comment by dsteurer** [comment]
I have added `build` and a `teardown` that implement more general versions of `init` and `halt!`.
Those functions should also be flexible enough to add suspend / resume functionality (later on or as a separate library). For that flexibility I thought it would be better to have `teardown` return the updated system map instead of `nil` like `halt!`. The definition of `halt!` in terms of `teardown` is now slightly awkward as it just returns `nil` after updating the system map using `halt-key!`. Another option would be to have `halt!` return the updated map but it would change the current behavior.

**Comment by weavejester** [comment]
It's usually not a good idea to expand the scope of a pull request. Can you remove the `teardown` code you added? It belongs in a separate PR, and I don't think the implementation is right.

File: src/integrant/core.clj [hunk 1]
```diff
@@ -48,12 +48,37 @@
 (defn- sort-keys [ks m]
   (sort (dep/topo-comparator (dependency-graph m)) ks))
 
+(defn- update-key [f m k]
```
Comment (weavejester) [inline src/integrant/core.clj]: Change the argument order to `[m k f]` instead, to match the argument order of `clojure.core/update`.

File: src/integrant/core.clj [hunk 2]
```diff
@@ -48,12 +48,37 @@
...
+  (update m k (comp (partial f k) (partial expand-key m))))
+
+(defn build
+  "Build system map from config map using f. Keys are traversed in
```
Comment (weavejester) [inline src/integrant/core.clj]: Docstring suggestion:

```clojure
  "Apply a function f to each key value pair in a configuration map. Keys are
  traversed in dependency order, and any references in the value expanded. The
  function should take two arguments, a key and value, and return a new value."

File: src/integrant/core.clj [hunk 3]
```diff
@@ -48,12 +48,37 @@
...
+  {:pre [(map? config)]}
+  (when-let [refs (seq (missing-refs config))]
+    (throw (ex-info (str "Missing definitions for refs: " (str/join ", " refs))
+                    {:reason ::missing-refs
+                     :config config
+                     :missing-refs refs})))
+  (-> (reduce (partial update-key f) config (sort-keys keys config))
+      (with-meta {::origin config})))
+
+(defn teardown
```
Comment (weavejester) [inline src/integrant/core.clj]: As mentioned in my earlier comment, let's move this to a separate PR.

File: test/integrant/core_test.clj [hunk 4]
```diff
@@ -46,3 +46,41 @@
   (is (thrown-with-msg? clojure.lang.ExceptionInfo
                         #"Missing definitions for refs: :integrant.core-test/b"
                         (ig/init {::a (ig/ref ::b)}))))
+
```
Comment (weavejester) [inline test/integrant/core_test.clj]: Double space should be single space.

File: test/integrant/core_test.clj [hunk 5]
```diff
@@ -46,3 +46,41 @@
...
+(deftest build-test
+  (is (= [{:a [:build :a [:build :b 1]]
+           :b [:build :b 1]}
+          [[:build :b 1]
+           [:build :a [:build :b 1]]]]
+         (build-log
+          {:a (ig/ref :b)
+           :b 1}))))
+
+(defn teardown-log [system]
```
Comment (weavejester) [inline test/integrant/core_test.clj]: Remember to remove teardown tests as well.

File: src/integrant/core.clj [hunk 6]
```diff
@@ -48,12 +48,29 @@
...
+              (sort-keys keys config))
+      (with-meta {::origin config})))
+
 (defn expand
   "Replace all refs with the values they correspond to."
   [config]
-  (reduce (fn [m k] (update m k (partial expand-key m)))
-          config
-          (sort-keys (keys config) config)))
+  (build config (keys config) (comp second vector)))
```
Comment (weavejester) [inline src/integrant/core.clj]: Instead of `(comp second vector)`, which took me a moment to work out, perhaps instead `(fn [_ v] v)`? It's more concise, faster, and easier to see what it does at a glance.

**Comment by weavejester** [comment]
Could you also change the commit message to:

    Add build function for custom initialization

Then we'll be ready to merge. Thanks for your work on this!
